arch/risc-v/src/common/supervisor/riscv_context.S

```diff
@@ -0,0 +1,234 @@
+/****************************************************************************
+ * arch/risc-v/src/common/supervisor/riscv_context.S
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+.file "riscv_context.S"
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <arch/mode.h>
+
+#include "riscv_exception_macros.S"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Symbols
+ ****************************************************************************/
+
+    .globl  riscv_saveusercontext
+    .globl  riscv_fullcontextrestore
+    .globl  riscv_switchcontext
+    .globl  riscv_syscall_return
+
+/****************************************************************************
+ * Name: riscv_saveusercontext
+ *
+ * Description:
+ *   Save user context, partially destroys the caller's context
+ *
+ * C Function Prototype:
+ *   int riscv_saveusercontext(uintptr_t *saveregs);
+ *
+ * Input Parameters:
+ *   saveregs - Context to save
+ *
+ * Returned Value:
+ *   0 on context switch
+ *   1 on no context switch
+ *
+ * Assumptions:
+ *   Global interrupts disabled by the caller.
+ *
+ ****************************************************************************/
+
+.type riscv_saveusercontext, function
+
+riscv_saveusercontext:
```

Comment:
can we reuse the syscall implementation like this:
riscv_saveusercontext->sys_call1->riscv_syscall_dispatch->riscv_handle_syscall.
So the machine and supervisor mode share all most code base.

Reply:
I personally prefer this, it is a simple context save to 1 location (a0) and does not need to involve a system call. This is just faster, even though it duplicates a bit of code.

If duplicated code is to be removed at all expense then I guess it can be done. In that case it would be the same as simulating ecall but with a trampoline.

Reply:
> I personally prefer this, it is a simple context save to 1 location (a0) and does not need to involve a system call.

But, the most portion of code is to simulate ECALL which is same as riscv_syscall_dispatch. Once the simulation is done, riscv_handle_syscall is just a simple switch case statement.

> This is just faster, even though it duplicates a bit of code.
> 

Yes, it could save some instruction, but is very small. On the other hand, it duplicated a lot of code and make the code hard to maintain.

> If duplicated code is to be removed at all expense then I guess it can be done. In that case it would be the same as simulating ecall but with a trampoline.

Reply:
Done